### DISCUSSION TITLE
How to restore a backup?

### DISCUSSION THREAD

**Discussion by technickfreak2000** [OpenSlides/OpenSlides#6343] [Q&A]
Hello,

I have installed the version 3.4.X and configured a backup of the data directory with the WAL archive.
In my case the data directory is located at: /var/lib/docker/volumes/instance1_dbdata1

I have yet not found a guide on how to restore the backup. 

Does anyone have a clue? 

Thank you in advance!

**Comment by ostcar** [comment]
All relevant data is stored in a postgres database. So you can follow any guide to backup and restore a postgres database.

I would recommend not to backup the raw-postgres-files but to create an sql-dump.

**Reply by technickfreak2000** [reply]
Thanks for your reply. 

Unfortunately, i am stuck again.

I have used this command to create an sql-dump: `docker exec -t instance1_pgnode1_1 pg_dumpall -c -U postgres > dump.sql`
and got the error: pg_dumpall: could not connect to database "template1": FATAL:  Peer authentication failed for user  "postgres"

I also tried changing the pg_hba.conf, although I am, from my understanding, connecting locally. So peer authentication should work...

Do you have an idea on how I am able to authenticate with the DB?

**Reply by ostcar** [reply]
Your command uses the wrong credentials. OpenSlides uses the postgres-user `openslides` with the password `openslides`. See https://github.com/OpenSlides/OpenSlides/blob/c03583989235b0fa9080fff2908304539514259c/docker/docker-compose.dev.yml#L27-L30

I have not tested it, but it is probably: `docker exec -t instance1_pgnode1_1 pg_dumpall -c -U openslides > dump.sql` 

Personally, I am more of a fan of `pg_dump DATABASE` then `pg_dumpall`. I don't like the "all-in-one" file. But it this case, it should not make a difference.

**Reply by ostcar** [reply]
This answer is for OpenSlides 3.4. For OpenSlides 4.0 see this issue: https://github.com/OpenSlides/openslides-manage-service/issues/143

**Reply by technickfreak2000** [reply]
I stumbled on this before. But I am getting the same result -> `pg_dumpall: could not connect to database "template1": FATAL:  Peer authentication failed for user "openslides"`

Furtheremore, I've tried the command inside the container:
`root@user:~# docker exec -t -i instance11_pgnode1_1 bash`
`root@450137d95ca2:/var/lib/postgresql# pg_dumpall -c -U openslides`
`pg_dumpall: could not connect to database "template1": FATAL:  Peer authentication failed for user "openslides"`
(No password prompt, nothing, just straight up the error...)

Maybe this could help identify my problem:
I have installed OpenSlides as documented with a self signed Certificate -> https://github.com/OpenSlides/OpenSlides/tree/stable/3.4.x
In the docker/.env file I've added the instance domain (in this case the IP), URL scheme (https) and ports (80 /443).
After generating the docker-compose.yml `( set -a; source .env; m4 docker-compose.yml.m4 ) > docker-compose.yml` I've only changed some proxy settings. -> [docker-compose.yml.txt](https://github.com/OpenSlides/OpenSlides/files/7972504/docker-compose.yml.txt)

I use this command to start the containers: `docker-compose -p instanceXX up -d`

OpenSlides itself works like a charm, so it is able to communicate with Postgres. What am I / are we just missing?

**Reply by ostcar** [reply]
I am not a postgres expert. But I think the peer authorization error means, that postgres uses the current Linux user name as username.

Maybe you have to change the pg_hba.conf file as described here

https://stackoverflow.com/a/18664239/4493241

Or you have to create a user called "root" in postgres and give him access to the openslides database.

Or you have to create a Linux user "openslides".

@peb-adr do you know how to create an SQL dump in OpenSlides 3.4?

**Reply by peb-adr** [reply]
@technickfreak2000 
peer auth is disabled for pg-user 'openslides'. You can verify it in `/etc/postgresql/11/main/pg_hba.conf` within the container.
What you probably want to do is peer authenticate as 'postgres' user.
Your command would look like this:
`docker exec -tu postgres instance1_pgnode1_1 pg_dumpall -c > dump.sql`
Mind the `-u postgres` flag for docker exec, which will log into the container as postgres. Then the peer auth succeeds.
Alternatively you could do password auth as 'openslides' by connecting through the tcp socket instead of unix socket by adding `-h 127.0.0.1` to the dump cmd. This requires an interactive shell however for entering the password

Also as ostcar mentioned I would go for only the db backup by doing
`docker exec -tu postgres instance1_pgnode1_1 pg_dump -c openslides > dump.sql`
But yeah shouldn't make much of a difference in your case.

**Reply by peb-adr** [reply]
To be clear, when peer authenticating the linux user and postgres user must match.
thats why merely adding `-U postgres` didn't work. You were still root, thus "peer auth fails for postgres".
